### PR TITLE
Porting libtokencap to Darwin.

### PR DESCRIPTION
**Pull request by devnexen**
Reading only main addresses and read only's.

### CODE REVIEW

**Comment by vanhauser-thc** [comment]
@devnexen : thanks!
@hexcoder- : how about porting this to OpenBSD at the same time?
(or even FreeBSD/NetBSD if one of you two is able to (system availabilty and time))

**Comment by devnexen** [comment]
I can help of any BSD if necessary.

**Comment by hexcoder-** [comment]
Great, thanks!
Yes, let us compare *BSDs APIs and integrate all of them (if possible).
I am able to run NetBSD, FreeBSD, OpenBSD in VirtualBox and Darwin on a Mac mini, but need to investigate the APIs before.

**Comment by devnexen** [comment]
@hexcoder since you seem to like OpenBSD maybe you would like to have a take on ? Did the FreeBSD implementation maybe we can merge it it is ok or still work from this branch ?

File: libtokencap/libtokencap.so.c [hunk 1]
```diff
@@ -102,6 +108,48 @@ static void __tokencap_load_mappings(void) {
...
+     struct kinfo_vmentry *region = (struct kinfo_vmentry *)low;
+     size_t size = region->kve_structsize;
+
+     if (size == 0) break;
+
+     /* We go through the whole mapping of the process and track read-only addresses */
+     if ((region->kve_protection & KVME_PROT_READ) &&
+	 !(region->kve_protection & KVME_PROT_WRITE)) {
+          __tokencap_ro[__tokencap_ro_cnt].st = (void *)low;
+          __tokencap_ro[__tokencap_ro_cnt].en = (void *)high;
```
Comment (hexcoder-) [inline libtokencap/libtokencap.so.c]: should be
`__tokencap_ro[__tokencap_ro_cnt].en = (void *)(low +size);`

**Comment by hexcoder-** [comment]
It checked the Darwin part and it is ok. Thanks very much. I enabled the corresponding test in the script for `make tests`. FreeBSD has problems yet, since the test does not succeed.
It is ok to work here for me.

**Comment by hexcoder-** [comment]
Test in subdirectory `test` with
```
cc -o test-compcov test-compcov.c
AFL_TOKEN_FILE=token.out LD_PRELOAD=../libtokencap.so ./test-compcov foobar
grep BUGMENOT token.out
```

File: libtokencap/libtokencap.so.c (hunk 1 above)
Comment (devnexen) [reply libtokencap/libtokencap.so.c]: ah yes in fact even the line just above is wrong too ... let me fix this

**Comment by hexcoder-** [comment]
Bingo!

**Comment by hexcoder-** [comment]
ok, can investigate NetBSD and OpenBSD tomorrow.

**Comment by vanhauser-thc** [comment]
good job guys!

**Comment by hexcoder-** [comment]
@devnexen: Could you help out with NetBSD? It partly works, but seems to grab invalid addresses,  which later lead to seg faults :-(

**Comment by devnexen** [comment]
Try to refactor for NetBSD case instead as :
`
for (I = 0; I<len; i++) {
... getting region as region[I]..
}
`
or maybe mmap flags to be changed or making a simple malloc for NetBSD.
Hope it helps.

**Comment by hexcoder-** [comment]
ok, I ask at the NetBSD mailing list then.

**Comment by devnexen** [comment]
might help in case but I took little time to look at it and in fact the segfault occurs when dumping the tokens. When I redirect to say stderr, it does not segfault.

**Comment by hexcoder-** [comment]
Thanks, that is interesting. So the iteration might be ok, then.

**Pull request merged**